Configurable Index Type

liran-funaro opened the pull request:
Fixes #10321.

See  #10321 for details.


<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist above are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

<hr>

##### Key changed/added classes in this PR
* `AppendableIndexBuilder`
* `AppendableIndexSpec`
 * `TuningConfig`
 * `AppenderatorConfig`

Eshcar commented on processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryHelper.java:
```diff
@@ -117,24 +120,24 @@ public String apply(DimensionSpec input)
+      indexBuilder = new OnheapIncrementalIndex.Builder();
     }
 
+    index = indexBuilder
```
great. reducing duplication

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -319,129 +316,6 @@ protected IncrementalIndex(
     }
   }
 
-  public static class Builder
```
so this static class has now changed to an abstract class with different concrete builder classes - nice

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -129,15 +136,11 @@ public FactsHolder getFacts()
-      } else {
-        aggOffsetInBuffer[i] = aggOffsetInBuffer[i - 1] + metrics[i - 1].getMaxIntermediateSizeWithNulls();
-      }
+      aggOffsetInBuffer[i] = aggsCurOffsetInBuffer;
```
nice - simpler

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -129,15 +136,11 @@ public FactsHolder getFacts()
 
-    aggsTotalSize = aggOffsetInBuffer[metrics.length - 1] + metrics[metrics.length
-                                                                    - 1].getMaxIntermediateSizeWithNulls();
+    aggsTotalSize = aggsCurOffsetInBuffer;
```
why is this correct?

liran-funaro replied:
After the last aggregator, the next offset in the buffer is the total size of the buffer.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -346,4 +349,99 @@ public void close()
+    }
+
+    @Override
+    protected OffheapIncrementalIndex buildInner()
```
this method should return IncrementalIndex
Is this legal to change the signature ?

liran-funaro replied:
It is legal for an implementation to increase the requirement, but not to lower it.
As long as it conforms with the interface's contract.

Eshcar commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -434,4 +436,39 @@ public ColumnCapabilities getColumnCapabilities(String columnName)
+  public static class Builder extends AppendableIndexBuilder
+  {
+    @Override
+    protected OnheapIncrementalIndex buildInner()
```
likewise - diff signature than parent class

Eshcar reviewed (approved):
This PR is bloated with similar changes due to refactoring 
specifically the change from buildOnHeap() to build() - it affects more than 100 files 
maybe there is a way to avoid it or have it scoped in a separate PR so that it would be easier to see and review exactly the changes that add the configuration capability

I added some comments suggesting how to further reduce the number of unnecessary changes (like null params, etc.) 

Other than that looks good to me

a2l007 reviewed (commented):
+1 for the extensible design that makes it easier to use new incremental index types. It would be useful to have integration tests for this which can test the appendableIndexSpec property within tuningConfig.

liran-funaro commented on processing/src/main/java/org/apache/druid/segment/incremental/AppendableIndexBuilder.java:
```diff
@@ -0,0 +1,128 @@
+    }
+  }
+
+  public final IncrementalIndex build()
```
You're right, it would be more elegant without this `buildInner()` method.
But I was trying to avoid code duplication: having to call `validate()` and `log.info()` in each `build()` implementation.
It is marked as `final` exactly to enforce this behavior.
I could rename it to something else, `instantiate()` for example.
What do you think?

liran-funaro replied:
Changed to DEBUG.

a2l007 commented on processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -346,4 +349,99 @@ public void close()
+          concurrentEventAdd,
+          sortFacts,
+          maxRowCount,
+          Objects.requireNonNull(bufferPool, "bufferPool is null")
```
Isn't validate() already checking this?

liran-funaro replied:
It does, but without this, Intellij warns about "Argument might be null".

a2l007 commented on processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:
```diff
@@ -346,4 +349,99 @@ public void close()
+
+    @JsonCreator
+    public Spec(
+        final @JsonProperty("bufferSize") @Nullable Integer bufferSize,
+        final @JsonProperty("cacheSize") @Nullable Integer cacheSize
+    )
+    {
+      this.bufferSize = bufferSize != null && bufferSize > 0 ? bufferSize : DEFAULT_BUFFER_SIZE;
+      this.cacheSize = cacheSize != null && cacheSize > this.bufferSize ? cacheSize : DEFAULT_CACHE_SIZE;
+      this.bufferPool = new StupidPool<>(
+          "Offheap incremental-index buffer pool",
+          this,
+          0,
+          this.cacheSize / this.bufferSize
```
The changes in this class apart from conforming to the AppendableIndexSpec interface might be out of scope of this PR and it might require its own additional tests. Would it be make sense to propose these changes in an incremental PR once the AppendableIndexSpec changes are merged?

liran-funaro replied:
Yes. That is possible. I will separate this part into a PR that will follow this one.

liran-funaro replied:
Removed

liran-funaro commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -319,126 +316,62 @@ protected IncrementalIndex(
+      return (Builder) super.setMaxBytesInMemory(maxBytesInMemory);
     }
 
     public OnheapIncrementalIndex buildOnheap()
```
I leave it on to avoid changing 100+ lines of code.
This can be removed later in a refactor PR.
Do you think I should remove it in this PR?

liran-funaro replied:
After reading your comment about the off-heap builder, I think it is inevitable to leave this builder and add back the `buildOffheap()` method as well.

liran-funaro replied:
Done. See #10494

liran-funaro commented on processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryHelper.java:
```diff
@@ -117,24 +120,24 @@ public String apply(DimensionSpec input)
         .build();
 
+
+    AppendableIndexBuilder indexBuilder;
```
Yes.

liran-funaro commented on server/src/main/java/org/apache/druid/segment/indexing/TuningConfig.java:
```diff
@@ -32,11 +35,43 @@
-  long DEFAULT_MAX_BYTES_IN_MEMORY = JvmUtils.getRuntimeInfo().getMaxHeapSizeBytes() / 6;
+
+  /**
+   * The inceremental index implementation to use
```
Nice catch

liran-funaro replied:
Fixed

liran-funaro commented on server/src/main/java/org/apache/druid/segment/indexing/TuningConfig.java:
```diff
@@ -32,11 +35,43 @@
+    }
+  }
+
+  PartitionsSpec getPartitionsSpec();
+
+  IndexSpec getIndexSpec();
+
+  IndexSpec getIndexSpecForIntermediatePersists();
```
We don't have to. I was wondering why they are in `AppenderatorConfig` in the first place.
The only reasoning for why `AppenderatorConfig` exists is because `HadoopTuningConfig` doesn't share all of the `TuningConfig` API with the rest of the "appenderators".
So I think it is best that as much of the common API as possible will be in `TuningConfig`.
And these methods are indeed in common.

liran-funaro replied:
> We should look at incrementally refactoring HadoopTuningConfig.getRowFlushBoundary into getMaxRowsInMemory so that it could be moved into TuningConfig as well.

I opened (earlier) #10478 for that. It is short.
If you prefer, I can apply your comments from this PR to #10478 and merge it before this one.
It address both `HadoopTuningConfig.getRowFlushBoundary` and the `TuningConfig` interface refactoring.

liran-funaro replied:
Hi @jihoonson. Can you chime in to resolve this issue?

jihoonson replied:
Oops, sorry. I missed the previous ping. 

> I was wondering why they are in `AppenderatorConfig` in the first place.

`indexSpecForIntermediatePersists` was added in #7919. I'm not aware of the exact reason, but it seems not bad to have it only in `AppenderatorConfig` because that parameter is coupled with how `Appenderator` spills intermediate segments. Hadoop task is sort of special. I think it was implemented before we added `Appenderator` (or at least before `AppenderatorDriver`), and we could implement `Appenderator` in a more structured way based on the lessons we learned from Hadoop task. So, it works similar to other tasks using `Appenderator`, but is not exactly same. 

Regarding the interface change, even though we currently have only the tasks which have the similar spilling mechanism, but it might not be true in the future. So, I would say it would be better to not modify the interface unless you have to.

liran-funaro replied:
Thanks, @jihoonson. I will revert this API change.

liran-funaro commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -1262,6 +1267,7 @@ private IndexTuningConfig(
+      this.appendableIndexSpec = appendableIndexSpec == null ? DEFAULT_APPENDABLE_INDEX : appendableIndexSpec;
       this.maxRowsInMemory = maxRowsInMemory == null ? TuningConfig.DEFAULT_MAX_ROWS_IN_MEMORY : maxRowsInMemory;
       // initializing this to 0, it will be lazily initialized to a value
       // @see server.src.main.java.org.apache.druid.segment.indexing.TuningConfigs#getMaxBytesInMemoryOrDefault(long)
```
Fixed

liran-funaro commented on server/src/main/java/org/apache/druid/segment/indexing/RealtimeTuningConfig.java:
```diff
@@ -132,6 +136,7 @@ public RealtimeTuningConfig(
+    this.appendableIndexSpec = appendableIndexSpec == null ? DEFAULT_APPENDABLE_INDEX : appendableIndexSpec;
     this.maxRowsInMemory = maxRowsInMemory == null ? DEFAULT_MAX_ROWS_IN_MEMORY : maxRowsInMemory;
     // initializing this to 0, it will be lazily initialized to a value
     // @see server.src.main.java.org.apache.druid.segment.indexing.TuningConfigs#getMaxBytesInMemoryOrDefault(long)
```
Fixed

liran-funaro commented on processing/src/main/java/org/apache/druid/segment/incremental/AppendableIndexSpec.java:
```diff
@@ -17,25 +17,16 @@
+import org.apache.druid.guice.annotations.UnstableApi;
+
+@UnstableApi
+public interface AppendableIndexSpec
```
Added

liran-funaro commented on processing/src/main/java/org/apache/druid/jackson/AppendableIndexModule.java:
```diff
@@ -0,0 +1,46 @@
+import org.apache.druid.segment.incremental.OffheapIncrementalIndex;
+import org.apache.druid.segment.incremental.OnheapIncrementalIndex;
+
+public class AppendableIndexModule extends SimpleModule
```
Added. See `AppendableIndexSpecTest`.

liran-funaro replied:
This is tested as part of the separate `TuningConfig` tests:
* `HadoopTuningConfigTest`
* `ParallelIndexTuningConfigTest`
* `KafkaIndexTuningConfigTest`
* `KafkaSupervisorTuningConfigTest`
* `KinesisIndexTaskTuningConfigTest`
* `KinesisSupervisorTuningConfigTest`
* `RealtimeTuningConfigTest`

liran-funaro commented on extensions-contrib/materialized-view-maintenance/src/test/java/org/apache/druid/indexing/materializedview/MaterializedViewSupervisorTest.java:
```diff
@@ -317,6 +318,32 @@ public void testCheckSegmentsAndSubmitTasks() throws IOException
 
   }
 
+  @Test
```
Added

liran-funaro commented:
Thanks, @a2l007 
I addressed all your comments in the codes, or with follow-up questions.
Regarding the integration tests, I added tests to all the `TuningConfig` tests that validate the `AppendableIndexSpec` within `TuningConfig`.
I think it is sufficient. Do you agree?

a2l007 reviewed (commented):
Thanks, I've left a couple more comments. 
Since we're introducing a new property within tuningConfig, I feel it would be useful to have a couple of integration tests with this property in the ingestion spec. However, I would be okay without adding them if the remaining community reviewers are fine without it.

Could you please fix the CI failures as well? Looks like there is coverage failure.

liran-funaro commented on indexing-hadoop/src/main/java/org/apache/druid/indexer/HadoopTuningConfig.java:
```diff
@@ -140,6 +144,7 @@ public HadoopTuningConfig(
     this.rowFlushBoundary = maxRowsInMemory == null ? maxRowsInMemoryCOMPAT == null
                                                       ? DEFAULT_ROW_FLUSH_BOUNDARY
                                                       : maxRowsInMemoryCOMPAT : maxRowsInMemory;
+    this.appendableIndexSpec = appendableIndexSpec == null ? DEFAULT_APPENDABLE_INDEX : appendableIndexSpec;
```
I'm unsure why this line needs to be documented. It is no different from any of the other tuning configurations in this constructor.
`AppendableIndexSpec` is documented, so it is self-explanatory.

liran-funaro commented on extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTuningConfigTest.java:
```diff
@@ -115,6 +115,7 @@ public void testSerdeWithNonDefaults() throws Exception
   public void testConvert()
   {
     KafkaSupervisorTuningConfig original = new KafkaSupervisorTuningConfig(
+        null,
```
I think it is best to avoid a new constructor here.
Otherwise, each new parameter will incur a new constructor.

liran-funaro commented on extensions-core/kinesis-indexing-service/src/test/java/org/apache/druid/indexing/kinesis/KinesisIndexTaskTest.java:
```diff
@@ -2739,6 +2739,7 @@ private KinesisIndexTask createTask(
     boolean resetOffsetAutomatically = false;
     int maxRowsInMemory = 1000;
     final KinesisIndexTaskTuningConfig tuningConfig = new KinesisIndexTaskTuningConfig(
+        null,
```
See above.

liran-funaro commented on indexing-service/src/main/java/org/apache/druid/indexing/common/index/RealtimeAppenderatorTuningConfig.java:
```diff
@@ -93,6 +96,7 @@ public RealtimeAppenderatorTuningConfig(
       @JsonProperty("maxSavedParseExceptions") @Nullable Integer maxSavedParseExceptions
   )
   {
+    this.appendableIndexSpec = appendableIndexSpec == null ? DEFAULT_APPENDABLE_INDEX : appendableIndexSpec;
```
I agree. But that will require a refactor to unify some of the `TuningConfig` implementations to one common (abstract?) implementation.
This can be done before or after this PR.

liran-funaro commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -1262,6 +1267,7 @@ private IndexTuningConfig(
         @Nullable Integer maxSavedParseExceptions
     )
     {
+      this.appendableIndexSpec = appendableIndexSpec == null ? DEFAULT_APPENDABLE_INDEX : appendableIndexSpec;
```
See above.

liran-funaro commented on indexing-hadoop/src/main/java/org/apache/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -302,11 +301,11 @@ private static IncrementalIndex makeIncrementalIndex(
         .build();
 
-    IncrementalIndex newIndex = new IncrementalIndex.Builder()
+    IncrementalIndex newIndex = tuningConfig.getAppendableIndexSpec().builder()
```
Done

liran-funaro commented:
> Could you please fix the CI failures as well? Looks like there is coverage failure.

The coverage gap comes from `ParallelIndexTuningConfig.equals()` and it was there before this PR. I just added another comparison.
It covers 10/20 branches, but it will require writing 10 additional unique tests just to cover it.

liran-funaro commented:
@a2l007 I addressed all your comments. Is there anything else I need to do to move this PR forward?

a2l007 commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskTuningConfig.java:
```diff
@@ -281,7 +292,8 @@ public boolean equals(Object o)
     SeekableStreamIndexTaskTuningConfig that = (SeekableStreamIndexTaskTuningConfig) o;
-    return maxRowsInMemory == that.maxRowsInMemory &&
+    return Objects.equals(appendableIndexSpec, that.appendableIndexSpec) &&
+           maxRowsInMemory == that.maxRowsInMemory &&
```
Could you please write an EqualsVerifier test for this. That should take care of the travis failures.

liran-funaro replied:
@a2l007 Thanks for the tip. I wasn't aware of `EqualsVerifier`.
I added it and now it passes the tests but fails some integration tests that passed before.
I suspect there is an issue with the integration tests because sometimes the master branch also fails in these.

a2l007 reviewed (commented):
Sorry for the delay. Just one additional comment, which should help with fixing the travis failures.

a2l007 reviewed (approved):
LGTM. Thanks!

liran-funaro commented:
@a2l007 Thanks!

liran-funaro commented:
@jihoonson Do you have any more comments or can we proceed to merge this PR?

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -1262,9 +1267,10 @@ private IndexTuningConfig(
       this.maxRowsInMemory = maxRowsInMemory == null ? TuningConfig.DEFAULT_MAX_ROWS_IN_MEMORY : maxRowsInMemory;
-      // initializing this to 0, it will be lazily initialized to a value
-      // @see server.src.main.java.org.apache.druid.segment.indexing.TuningConfigs#getMaxBytesInMemoryOrDefault(long)
+      /** initializing this to 0, it will be lazily initialized to a value
+       * @see #getMaxBytesInMemoryOrDefault() */
```
nit: we don't use multi-line comments.

liran-funaro replied:
Fixed.

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskTuningConfig.java:
```diff
@@ -84,10 +87,11 @@ public SeekableStreamIndexTaskTuningConfig(
     this.partitionsSpec = new DynamicPartitionsSpec(maxRowsPerSegment, maxTotalRows);
-    // initializing this to 0, it will be lazily initialized to a value
-    // @see server.src.main.java.org.apache.druid.segment.indexing.TuningConfigs#getMaxBytesInMemoryOrDefault(long)
+    /** initializing this to 0, it will be lazily initialized to a value
+     * @see #getMaxBytesInMemoryOrDefault() */
```
Same here.

liran-funaro replied:
Fixed.

jihoonson commented on docs/ingestion/index.md:
```diff
@@ -737,3 +741,11 @@ The `indexSpec` object can include the following properties:
+|-----|-----------|-------|
+|type|Each in-memory index has its own tuning type code. You must specify the type code that matches your in-memory index. Common options are `onheap`, and `offheap`.|`onheap`|
+
+Beyond these properties, each in-memory index has its own specific tuning properties.
```
Do you think if users want to change this config ever for now? I remember @gianm mentioned that the offheap incremental index has some performance issue, which is why we don't use it for indexing. If you think this knob is useful for users, please add more details how each index type is different and when it's recommended to use what. Otherwise, I would suggest not documenting this know at least for now.

liran-funaro replied:
Actually, by @a2l007 suggestion, I deferred the support for offheap ingestion to a different PR.
So this documentation should not include the `offheap` option anyway.
I'll removed this documentation for now.

jihoonson reviewed (commented):
Hi @liran-funaro, I haven't reviewed line by line, but left some comments especially for documentation.

liran-funaro commented:
@jihoonson Thanks. I address all your comments.
Please let me know if you see any other required modifications, or that we could proceed to merge this PR.

jihoonson reviewed (approved):
LGTM. +1 after CI. Thanks @liran-funaro!

liran-funaro commented:
Passed CI.
Thanks @jihoonson and @a2l007

liran-funaro commented:
@jihoonson Do we need more approval before merge?

jihoonson commented:
@liran-funaro we usually wait for some time before merging a PR for other committers if there is anyone who wants to review as well. I will merge this PR today unless there is someone else who has more comments.

jihoonson commented:
@liran-funaro merged. Thanks!

exherb commented:
Looks this feature has been removed? I try to index kafka stream with off heap but got ignored.
